File: ruby/text/paraphrase_xml.rb
# encoding: UTF-8
#
# Example script to issue a paraphrase.xml request
#
require 'rubygems'
require 'uri'
require 'net/http'
require 'digest/md5'
require 'openssl'
require 'base64'
require 'time'
require 'zlib'
require 'rexml/document'

# Set your environment variables to the keys obtained from https://www.idilia.com/developer/my-projects
accessKey = ENV['IDILIA_ACCESS_KEY'] || ""
privateKey = ENV['IDILIA_PRIVATE_KEY'] || ""
raise "You have to set environment variables IDILIA_ACCESS_KEY and IDILIA_PRIVATE_KEY" if accessKey.empty? or privateKey.empty?

# The queries that we'll process
texts=["porch lights", "car engine problems diagnosing", "nikon digital camera"]
textMime="text/query; charset=iso-8859-1"

# Open up a connection that we can re-use for several requests
uri = URI('http://api.idilia.com/1/text/paraphrase.xml')
Net::HTTP.start(uri.host, uri.port) do | http |

  texts.each do | text |
    
    # Compute the authorization
    date = Time.now.httpdate()
    md5 = Base64.encode64(Digest::MD5.digest(text)).chomp
    toSign=date + '-' + uri.host + '-' + uri.path + '-' + md5
    signature=Base64.encode64(OpenSSL::HMAC.digest('sha256', privateKey, toSign)).chomp
    authorization = "IDILIA " + accessKey + ":" + signature
  
    # Create the HTTP request with the headers and parameters
    request = Net::HTTP::Post.new(uri.path, {
      'Accept-Encoding' => 'gzip',
      'Authorization' => authorization,
      'Date' => date,
      'Host' => uri.host
      })
    request.set_form_data({
        'text' => text,
        'textMime' => textMime,
        'maxCount' => 10,
        'minWeight' => 0.5,
        'paraphrasingRecipe' => 'productSearch',
        'requestId' => 'my-request'
      })
    request.set_content_type('application/x-www-form-urlencoded; charset=UTF-8')
  
    # Get the response and decode it
    http.read_timeout = 3600 + 60; # slightly exceeds default value for parameter "timeout"
    httpResponse = http.request(request)
    body = httpResponse.body
    body = Zlib::Inflate.new(16+Zlib::MAX_WBITS).inflate(body) if httpResponse['Content-Encoding'] == 'gzip'
  
    # Parse the response
    case httpResponse
    when Net::HTTPOK
      doc = REXML::Document.new(body)
      ccfmp = doc.root.elements["queryConf/confCorrectFineMostProbable"].text
      puts "Paraphrases received for query: [#{text}] (overall conf: #{ccfmp})"
      doc.elements.each("//paraphrase") do | p |
        puts "  [#{p.elements["surface"].text}] with weight #{p.elements["weight"].text}"
      end
      puts
  
    else
      puts body
      raise "Some unexpected error occured when processing query [#{text}]"
    end
  end
end

File: python/text/paraphrase_xml.py
#
# Example program for issuing paraphrase.xml request
# processed immediately on response.
#

import urllib
import os
import httplib
import urlparse
import hashlib
import hmac, base64
import zlib
import xml.etree.cElementTree as xml
from email.utils import formatdate
from datetime import datetime
from time import mktime

URL = "http://api.idilia.com/1/text/paraphrase.xml"
ACCESS_KEY = os.environ.get("IDILIA_ACCESS_KEY")
PRIVATE_KEY = os.environ.get("IDILIA_PRIVATE_KEY")
if (ACCESS_KEY == None or PRIVATE_KEY == None or len(ACCESS_KEY) == 0 or len(PRIVATE_KEY) == 0):
  raise Exception("You have to set environment variables IDILIA_ACCESS_KEY and IDILIA_PRIVATE_KEY")


# Setup the HTTP connection. It can be re-used for multiple 
# consecutive requests as we use HTTP 1.1
url = urlparse.urlparse(URL)
httpConn = httplib.HTTPConnection(url.hostname, url.port, timeout=4000)

# Create the request params.
texts=["porch lights", "car engine problems diagnosing", "nikon digital camera"]

for text in texts:
  # Compute the authorization
  date = formatdate(timeval=mktime(datetime.utcnow().timetuple()), localtime=True, usegmt=True)
  md5 = base64.b64encode(hashlib.md5(text).digest())
  toSign = "%s-%s-%s-%s" % (date, url.hostname, url.path, md5)
  signature = str(base64.b64encode(hmac.new(PRIVATE_KEY, toSign, hashlib.sha256).digest()))
  
  # The request parameters
  params = {
    'requestId' : 'mytest',
    'text' : text,
    'maxCount' : 10,
    'minWeight' : 0.5,
    'paraphrasingRecipe' : 'productSearch',
    'textMime' : "text/query; charset=us-ascii"
  }
  
  # The HTTP headers that we use: Mostly for validation
  headers = {
      'Accept-Encoding': 'gzip',
      'Authorization' : "IDILIA %s:%s" % (ACCESS_KEY, signature),
      'Content-Type': "application/x-www-form-urlencoded; charset=UTF-8",
      'Date' : date,
      'Host' : url.hostname
  }
  
  # Send the POST request and send it
  httpConn.request("POST", url.path, urllib.urlencode(params), headers)
  
  # Get the response and decode it
  httpResponse = httpConn.getresponse();
  httpBody = httpResponse.read();
  if httpResponse.getheader("Content-Encoding") == "gzip":
    httpBody = zlib.decompress(httpBody, 16+zlib.MAX_WBITS)
  
  # First validation in case something went wrong in the transport layer
  if httpResponse.status != httplib.OK:
      raise Exception(httpBody, httpResponse.status, httpResponse.reason)
  
  # Application layer validation
  resp = xml.fromstring(httpBody)
  if resp.find('errorMsg') != None:
      raise Exception(resp.find('errorMsg').text, resp.find('status').text)
  
  # Display the senses found.
  ccfmp = resp.find('queryConf/confCorrectFineMostProbable').text
  print "Paraphrases received for query: [%s] (overall conf: %s)" % (text, ccfmp)
  for p in resp.findall(".//paraphrase"):
    print "  [%s] with weight %s" % (p.find("surface").text, p.find("weight").text)
  print
